File: sourcecode/1743.py
    import sys
    sys.setrecursionlimit(10000)

    n,m,k = map(int, input().split())

    matrix = [[0]*m for i in range(n)]
    visited = [[False]*m for i in range(n)]

    dx = [1,-1,0,0]
    dy = [0,0,1,-1]

    for i in range(k):
        r,c = map(int, input().split())
        matrix[r-1][c-1] = 1


    count = 1

    def dfs(x,y):
        global count
        visited[x][y] = True

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0<=nx<n and 0<=ny<m:
                if matrix[nx][ny]==1 and visited[nx][ny] == False:
                    count += 1
                    dfs(nx, ny)
        
        return count


    trash = []

    for i in range(n):
        for j in range(m):
            if matrix[i][j] == 1 and visited[i][j] == False:
                temp = dfs(i,j)
                trash.append(temp)
                count = 1

    print(max(trash))

File: sourcecode/1743.c
#include <stdio.h>

#define MAX 101

int N,M,K;
int matrix[MAX][MAX];
int visited[MAX][MAX];

int dx[] = {1,-1,0,0};
int dy[] = {0,0,1,-1};

int answer = 1;
int check = 0;

void init(){
    for(int i=0; i<MAX; i++){
        for(int j=0; j<MAX; j++){
            matrix[i][j] = 0;
            visited[i][j] = 0;
        }
    }
}

void dfs(int x, int y){
    visited[x][y] = 1;

    for(int i=0; i<4; i++){
        int nx = x + dx[i];
        int ny = y + dy[i];

        if (0<=nx && nx<N && 0<=ny && ny<M){
            if (matrix[nx][ny] == 1 && visited[nx][ny] == 0){
                visited[nx][ny] = 1;
                answer += 1;
                dfs(nx, ny);
            }
        }
    }

}

int main(void){
    scanf("%d %d %d", &N, &M, &K);
    init();


    for(int i=0; i<K; i++){
        int a;
        int b;
        scanf("%d %d", &a, &b);
        matrix[a-1][b-1] = 1;
    }

    // for(int i=0; i<N; i++){
    //     for(int j=0; j<M; j++){
    //         printf("%d ", matrix[i][j]);
    //     }
    //     printf("\n");
    // }

    for(int i=0; i<N; i++){
        for(int j=0; j<M; j++){
            if(matrix[i][j] == 1 && visited[i][j]==0){
                dfs(i,j);
                if(answer>check){
                    check = answer;
                }
                answer = 1;
            }
        }
    }

    // for(int i=0; i<N; i++){
    //     for(int j=0; j<M; j++){
    //         printf("%d ", matrix[i][j]);
    //     }
    //     printf("\n");
    // }

    printf("%d", check);

    return 0;
}